### ISSUE TITLE
depthai 2.7.2 led java coredump. But depthai 2.6.0 is fine.

### ISSUE THREAD

**Issue by wureka** [bytedeco/javacpp#509]
Environment:
- OS: Ubuntu 20.04 Desktop 64bit LTS
- JDK:  GraalVM JDK 21.1.0.r11-grl
- Micronaut 2.5.9
- JavaCPP: 1.5.6-SNAPSHOT
- OpenCV: 4.5.3
- Depthai: 2.7.2
- test code: [depthai-java-test](https://github.com/wureka/depthai-java-test)

Reproduce steps:

```
rm -rf ~/.m2/repository/org/bytedeco/*
rm -rf ~/.javacpp
cd ~/tmp/
git clone https://github.com/bytedeco/javacpp-presets
cd javacpp-presets
mvn install --projects openblas,opencv,depthai -Djavacpp.platform=linux-x86_64
cd ~/tmp/
git clone https://github.com/wureka/depthai-java-test
cd depthai-java-test
sh gradlew shadowJar
java -Djava.library.path=~/tmp/javacpp-presets/depthai/target/native/org/bytedeco/depthai/linux-x86_64:~/tmp/javacpp-presets/openblas/target/native/org/bytedeco/openblas/linux-x86_64:~/tmp/javacpp-presets/opencv/target/native/org/bytedeco/opencv/linux-x86_64 -jar build/libs/depthai-java-test-0.1-all.jar
```
The actual output is:

```
18:29:22.118 [main] INFO  i.m.context.env.DefaultEnvironment - Established active environments: [cli]
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007f3512c716e8, pid=704492, tid=704493
#
# JRE version: OpenJDK Runtime Environment GraalVM CE 21.1.0 (11.0.11+8) (build 11.0.11+8-jvmci-21.1-b05)
# Java VM: OpenJDK 64-Bit Server VM GraalVM CE 21.1.0 (11.0.11+8-jvmci-21.1-b05, mixed mode, sharing, tiered, jvmci, jvmci compiler, compressed oops, g1 gc, linux-amd64)
# Problematic frame:
# C  [libdepthai-core.so+0x1836e8]  tcpip_get_devices+0x148
#
# Core dump will be written. Default location: Core dumps may be processed with "/usr/share/apport/apport %p %s %c %d %P %E" (or dumping to /home/alexji/java-projects/depthai-projects/depthai-java-test/core.704492)
#
# An error report file with more information is saved as:
# /home/alexji/java-projects/depthai-projects/depthai-java-test/hs_err_pid704492.log
#
# If you would like to submit a bug report, please visit:
#   https://github.com/oracle/graal/issues
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#
Aborted (core dumped)

```

If I checkout below code.
![image](https://user-images.githubusercontent.com/9668764/126895144-aa60e6c9-111f-4723-8656-34086dad39fe.png)

And modify build.gradle
`implementation('org.bytedeco:depthai:2.6.0-1.5.6-SNAPSHOT')`

Re-do the above steps. the same code will work. The correct output is:

```
17:46:11.664 [main] INFO  i.m.context.env.DefaultEnvironment - Established active environments: [cli]
Detect 3 camera(s), USB speed: SUPER 
    Camera 0 is ready!
    Camera 1 is ready!
    Camera 2 is ready!

Process finished with exit code 0
```

**Comment by saudet** [comment]
I'm afraid there is nothing significant that changed with the presets between 2.6.0 and 2.7.2:
https://github.com/bytedeco/javacpp-presets/compare/1a619b0..2182762
This is an issue with DepthAI itself. Please report upstream: https://github.com/luxonis/depthai-core

**Comment by saudet** [comment]
BTW, please consider using OpenVINO directly instead, if it contains all the features that you need.
It's a more mature, thus more stable library than DepthAI. /cc @Grumpy141

**Comment by saudet** [comment]
Does this still happen with DepthAI 2.8.0?

**Comment by wureka** [comment]
> Does this still happen with DepthAI 2.8.0?
Yes. Actually, I doubt the problem only happens in Ubuntu 20.04. 
I use two new-installed and clean guest OS running in VMWare workstation 16  for test. One is Elementary OS 5 (based on Ubuntu 18.04 64bit LTS), and the other is Elementary OS 6 (based on Ubuntu 20.04 64bit LTS).
I built a python environment followed by [First steps with DepthAI](https://docs.luxonis.com/en/latest/pages/tutorials/first_steps/)
In Elementary OS 5, depthai 2.5.0, 2.6.0, 2.7.x, and 2.8.0 work fine. In Elementary OS6, however, only depthai 2.6.0 or below version works. I have collected some output of gdb followed by Luxois team's instruction and reported to Luxois team.

**Comment by saudet** [comment]
I've since upgraded the builds for the snapshots to Bionic, and according to @Grumpy141 it fixes the crashes on close for ARM, so could you try to take those binaries built on Bionic and try to run them as is on Focal? It's possible that DepthAI is only being tested on Bionic, and may not build properly on either Xenial or Focal, but binaries built on Bionic could still run correctly on Focal.

**Comment by wureka** [comment]
@saudet First, let you know what I had done in Last 2week.
1. Built javacpp-presets-1.5.6 in a newly installed Ubuntu 20.04 64bit in VMware workstation 16 and ran depthai sample with Java.
2. Built javacpp-presets-1.5.6 in my existed Ubuntu 20.04 64bit in my laptop and ran depthai sample with Java.

In case 1, depth-core 2.8.0 Java code worked. In case 2, howerer, depth-core 2.8.0 led Java coredump. I guess there could be some packages I installed in my laptop are not compatible with depth-core 2.8.0. I had reported this situation to Depthai team and followed their instructions for collecting some information of gdb to Depthai team.

Second, recently, depthai-core has upgraded to version 2.9.0. I guess that what you said the above problem has be resolevd should be in depthai-core 2.9.0.  But in javacpp-presets-1.5.7.SNAPSHOT, I found that the depthai still uses depthai-core 2.8.0. 

Anyway, I still gave a test on Javacpp-presets-1.5.7.SNAPSHOT last night but no luck. Same Java coredump.
After the Javacpp-presets-1.5.7.SNAPSHOT/depthai is changed to 2.9.0, I will try again.

**Comment by saudet** [comment]
2.9.0-1.5.7-SNAPSHOT is available, yes:
https://oss.sonatype.org/content/repositories/snapshots/org/bytedeco/depthai/2.9.0-1.5.7-SNAPSHOT/

**Comment by wureka** [comment]
> 2.9.0-1.5.7-SNAPSHOT is available, yes:
> https://oss.sonatype.org/content/repositories/snapshots/org/bytedeco/depthai/2.9.0-1.5.7-SNAPSHOT/

@saudet  Due to that I don't know how to set the JVM parameter java.library.path when using [https://oss.sonatype.org/content/repositories/snapshots](https://oss.sonatype.org/content/repositories/snapshots) as the Gradle's repository, I still clone [https://github.com/bytedeco/javacpp-presets](https://github.com/bytedeco/javacpp-presets) and built myself.
Anyway, with the same test as the above [https://github.com/bytedeco/javacpp/issues/509#issuecomment-898027594](https://github.com/bytedeco/javacpp/issues/509#issuecomment-898027594), again, depthai 2.9.0 still failed to work in my laptop. But it still works in the same OS in my guest machine of my vmware workstation 16 pro.

I tried Ubuntu 20.04, Linux Mint 20.2, Elementary OS 6, depthai-2.9.0 works without problem.

So, I believe that there should be some packages in my Ubuntu 20.04 Desktop in my laptop that are not compatilbe. But I don't know how to find the packages.
Forget it. Maybe it's a good time for me to reinstall Linux into my laptop. 
And I suggest this issue could be closed.

**Comment by saudet** [comment]
Is this still the case with DepthAI 2.11.1?

**Comment by wureka** [comment]
> Is this still the case with DepthAI 2.11.1?

No.  But it appears another problem.
Environment: Zorin OS 16 Pro (based on Ubuntu 20.04 64bit LTS)
JDK: OpenJDK 11.0.12

build.gradle:
```
plugins {
    id 'java'
    id 'java-library'
    id("com.github.johnrengelman.shadow") version "7.0.0"
}

ext {
    javacppVersion = "1.5.7-SNAPSHOT"
}
group 'org.example'
version '1.0'

repositories {
    maven { url 'https://oss.sonatype.org/content/repositories/snapshots/' }
    // mavenLocal()
    mavenCentral()
}

dependencies {
    implementation("org.bytedeco:openblas-platform:0.3.18-${javacppVersion}")
    implementation("org.bytedeco:opencv-platform:4.5.4-${javacppVersion}")
    implementation("org.bytedeco:depthai-platform:2.11.1-${javacppVersion}")
    
    testImplementation 'org.junit.jupiter:junit-jupiter-api:5.7.0'
    testRuntimeOnly 'org.junit.jupiter:junit-jupiter-engine:5.7.0'
}

test {
    useJUnitPlatform()
}

java {
    sourceCompatibility = JavaVersion.toVersion("11")
    targetCompatibility = JavaVersion.toVersion("11")
}
```

RgbPreview.java:

```
package org.example;

import org.bytedeco.depthai.*;
import org.bytedeco.javacpp.IntPointer;
import org.bytedeco.javacpp.Pointer;
import org.bytedeco.javacpp.PointerScope;
import org.bytedeco.opencv.opencv_core.Mat;

import static org.bytedeco.depthai.global.depthai.X_LINK_BOOTED;
import static org.bytedeco.opencv.global.opencv_highgui.imshow;
import static org.bytedeco.opencv.global.opencv_highgui.waitKey;

public class RgbPreview {
    static boolean INTERLEAVED = false;
    static int WIDTH = 416, HEIGHT = 416;

    static void tsleep(long tsec) {
        try {
            Thread.sleep(tsec);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }

    static void test1() {
        Pipeline pipeline  = new Pipeline();
        ColorCamera camRgb = pipeline.createColorCamera();
        XLinkOut xoutRgb   = pipeline.createXLinkOut();

        xoutRgb.setStreamName("rgb");
        camRgb.setResolution(ColorCameraProperties.SensorResolution.THE_1080_P);
        camRgb.setPreviewSize(WIDTH, HEIGHT);
        camRgb.setInterleaved(INTERLEAVED);
        camRgb.setColorOrder(ColorCameraProperties.ColorOrder.BGR);

        camRgb.preview().link(xoutRgb.input());



        Device device = new Device();

        System.out.println("MX ID:" + device.getMxId().getString());
        device.startPipeline(pipeline);
        IntPointer cameras = device.getConnectedCameras();
        System.out.println("Check point 00001----------------");

        // ------ ERROR BLOCK BEGINS: below leads to: ----------
        //        Stack trace (most recent call last) in thread 96998:
        //        #0    Object "[0x7fbab12927ad]", at 0x7fbab12927ad, in
        //        Segmentation fault (Address not mapped to object [(nil)])

        System.out.printf("Detect %s camera(s) \n", cameras);
        System.out.println("Check point 00002----------------");
        System.out.printf("USB speed: %s \n", device.getUsbSpeed());
        System.out.println("Check point 00003----------------");
        for (int i = 0; i < cameras.limit(); i++) {
            System.out.printf("    Camera %d is ready!\n", cameras.get(i));
        }
        System.out.println("Check point 00004----------------");
        // ------ ERROR BLOCK ENDS --------------------

        DataOutputQueue qRgb = device.getOutputQueue("rgb", 4, false);

        Mat frame;
        long time0 = 0, time1 = 0, accumConvertTime = 0;
        int count = 0;
        final int TOTAL_FRAME_COUNT = 1000;

        while (true) {
            try (PointerScope scope = new PointerScope()) {
                ImgFrame imgFrame = qRgb.getImgFrame();
                if (imgFrame != null && !imgFrame.isNull()) {
                    time0 = System.currentTimeMillis();
                    frame = imgFrame.getCvFrame();
                    time1 = System.currentTimeMillis();
                    accumConvertTime += (time1 - time0);
                    count++;
                    imshow("preview", frame);

                    if (count >=TOTAL_FRAME_COUNT) break;
                    int key = waitKey(1);
                    if (key == 'q' || key == 'Q') {
                        break;
                    }
                } else {
                    System.out.println("Not ImgFrame");
                }
            }

        }
        device.close();
        pipeline.close();
    }
    public static void main(String[] args) {
        test1();
    }
}

```

Output:

> MX ID:14442C1021********
> Stack trace (most recent call last) in thread 88413:
> #0    Object "[0x7f20b928ff2d]", at 0x7f20b928ff2d, in 
> Segmentation fault (Address not mapped to object [(nil)])

If I remove code between ERROR BLOCK BEGINS and ERROR BLOCK ENDS, the the above code can work. 

So, I have no idea which part causes the above error. JavaCPP ? or DepthAI 2.11.1 ?

**Comment by saudet** [comment]
There's nothing that changed in the presets that could trigger something like that. It's probably something in DepthAI. If you can narrow down the exact version that stopped working, I'm sure the authors upstream would appreciate the information.

**Comment by wureka** [comment]
> There's nothing that changed in the presets that could trigger something like that. It's probably something in DepthAI. If you can narrow down the exact version that stopped working, I'm sure the authors upstream would appreciate the information.

OK. I just gave it a test, and realized that the problem happened since depthai-2.11.0
depthai-2.10.0 works well.

**Comment by saudet** [comment]
The biggest differences between 2.10.0 and 2.11.0 seem to be BootloaderConfig, NetworkConfig, and UsbConfig: https://github.com/bytedeco/javacpp-presets/commit/fb4bba08164174377bb625bc9823b10114a6187c Maybe there's something we need to do with those?

**Comment by wureka** [comment]
Yes. I suggest the Luxonis's team can take some investigations. Could @saudet invite them ?

**Comment by saudet** [comment]
Invite them, why? Simply report the issue upstream. They should be able to reproduce this without Java.

**Comment by Grumpy141** [comment]
I have been running on 2.8.  I have not tried the newer versions.  I will try later this week when I get some time.  I am running java software that has been running continuously for weeks without issues.  I do turn the camera off when not actively using it.  

    On Tuesday, November 9, 2021, 11:48:36 PM MST, Samuel Audet ***@***.***> wrote:  
 
 


Invite them, why? Simply report the issue upstream. They should be able to reproduce this without Java.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by themarpe** [comment]
Hi @wureka 
Can you try building depthai in debug mode? The crash should provide much more information in that case, and could help exposing what the problem is (if it is crashing at depthai code and not later on somewhere else)

**Comment by wureka** [comment]
@themarpe  I wish I could. But I am not familiar with C++. Could anyone provide me some instruction about how to build depthai in debug mode in javacpp-presets  ?

**Comment by wureka** [comment]
A new test result:
If I run the above code in IntelliJ-IDEA 2021.2.3, the above error will appear.

However, if I execute below commands in terminal, the code will work

1. build the above code (`sh gradlew shadowJar`)
2. run the built jar file

**Comment by saudet** [comment]
> @themarpe I wish I could. But I am not familiar with C++. Could anyone provide me some instruction about how to build depthai in debug mode in javacpp-presets ?

Since it's using CMake, I'm guessing that we just need to replace `-DCMAKE_BUILD_TYPE=Release` with `-DCMAKE_BUILD_TYPE=Debug` in this file: https://github.com/bytedeco/javacpp-presets/blob/master/depthai/cppbuild.sh

**Comment by wureka** [comment]
> > @themarpe I wish I could. But I am not familiar with C++. Could anyone provide me some instruction about how to build depthai in debug mode in javacpp-presets ?
> 
> Since it's using CMake, I'm guessing that we just need to replace `-DCMAKE_BUILD_TYPE=Release` with `-DCMAKE_BUILD_TYPE=Debug` in this file: https://github.com/bytedeco/javacpp-presets/blob/master/depthai/cppbuild.sh

I use the above flags to re-build depthai-2.11.1 in javacpp-presets and test the above code in IntelliJ-IDEA 2021.2.3 with GraalVM-21.2.0.r11 again. The result is the same. But no extra information is dumpped.


```
> Task :RgbPreview.main()
MX ID:14442C102176E1****
Check point 00001----------------
Stack trace (most recent call last) in thread 23259:
#0    Object "[0x7fc8fd2848ad]", at 0x7fc8fd2848ad, in 
Segmentation fault (Address not mapped to object [(nil)])

> Task :RgbPreview.main() FAILED

Execution failed for task ':RgbPreview.main()'.
> Process 'command '/home/alexji/.sdkman/candidates/java/21.2.0.r11-grl/bin/java'' finished with non-zero exit value 139

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

```

**Comment by saudet** [comment]
You're not getting any hs_err_pidXXX.log files either?

**Comment by themarpe** [comment]
@wureka hmm, maybe try running under gdb in this case. The following should print the stack:
```
gdb java
run [name/of/the/java/program]
... (after crash)
bt
```

If compiling as debug didn't include any information there its possible that this comes from some other source

**Comment by saudet** [comment]
@wureka To make sure it's not something getting deallocated prematurely, try to set the "org.bytedeco.javacpp.nopointergc" system property to "true".

**Comment by wureka** [comment]
Today I tested one more time.
1. Make sure `-DCMAKE_BUILD_TYPE=Debug` is in the block of linux-x86 and linux-x86_64.
2. run 'mvn install --projects depthai -Djavacpp.platform=linux-x86_64`.
3. add below code :
```
    public static void main(String[] args) {
        System.setProperty("org.bytedeco.javacpp.nopointergc", "true");
        test1();
    }
```
4. config JDK to GraalVM-21.2.0.r11 (also including GraalVM-21.3.0.r11)
5. run the above test code in IntelliJ-IDEA.

Result:

```
> Task :RgbPreview.main()
MX ID:14442C102176E1D***
Check point 00001----------------
Stack trace (most recent call last) in thread 45518:
#0    Object "[0x7fbd6528582d]", at 0x7fbd6528582d, in 
Segmentation fault (Address not mapped to object [(nil)])

> Task :RgbPreview.main() FAILED

Execution failed for task ':RgbPreview.main()'.
> Process 'command '/home/alexji/.sdkman/candidates/java/21.2.0.r11-grl/bin/java'' finished with non-zero exit value 139

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

```

Again, if I config JDK to OpenJDK 11.0.12 in IntelliJ-IDEA and run the code in IntelliJ-IDEA, the code works.

```
10:52:42 a.m.: Executing ':RgbPreview.main()'...

> Task :compileJava UP-TO-DATE
> Task :processResources NO-SOURCE
> Task :classes UP-TO-DATE

> Task :RgbPreview.main()
MX ID:14442C102176E1D***
Check point 00001----------------
Detect org.bytedeco.javacpp.IntPointer[address=0x7ffabdb26b90,position=0,limit=3,capacity=3,deallocator=org.bytedeco.javacpp.Pointer$NativeDeallocator[ownerAddress=0x7ffabdb26b90,deallocatorAddress=0x7ff8a880c000]] camera(s) 
Check point 00002----------------
USB speed: SUPER 
Check point 00003----------------
    Camera 0 is ready!
    Camera 1 is ready!
    Camera 2 is ready!
Check point 00004----------------

Deprecated Gradle features were used in this build, making it incompatible with Gradle 8.0.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

See https://docs.gradle.org/7.1/userguide/command_line_interface.html#sec:command_line_warnings

BUILD SUCCESSFUL in 6s
2 actionable tasks: 1 executed, 1 up-to-date
10:52:49 a.m.: Execution finished ':RgbPreview.main()'.
```

**Comment by themarpe** [comment]
@wureka I'm not sure what the differences between OpenJDK and GraalVM are to cause this. Not that familiar with bindings as well. From core perspective, if you have happened to capture the stack using gdb, I could check that out, otherwise not sure what could go wrong in this case (apart from some sort of memory corruption)

**Comment by wureka** [comment]
@themarpe Well, I am not familiar with gdb. If you could provide me the operation steps, I can do that and feedback.
Anyway, I can accept using OpenJDK. After all, it is no difference to me.

**Comment by saudet** [comment]
@wureka Some possibly relevant bugs have been fixed in 2.13.1:
https://github.com/luxonis/depthai-core/releases
You might want to try out that version

**Comment by themarpe** [comment]
@wureka here - the following should do:
```
gdb java
run [name/of/the/java/program] [args if any...]
... (after crash)
bt
```

Change `java` to whatever is the main program that starts the crashing example (in your case that seems to be: `/home/alexji/.sdkman/candidates/java/21.2.0.r11-grl/bin/java`).

(The above basically forms a command line execution equivalent of eg: `> java myprogram [additional args]...` 


Also can you reproduce the issue with OpenJDK at all or is it GraalVM only?

**Comment by saudet** [comment]
FWIW, GraalVM bundles OpenJDK, it's essentially the same thing. There shouldn't be any difference. From DepthAI's point of view it's probably just memory corruption that happens to not cause any problems with a minor difference in the version of OpenJDK or how it was built.

@themarpe BTW, the JDK doesn't run the main() method on the first thread of the process. If this is known to cause problems for DepthAI, then you'll need to fix that for Java users.

**Comment by themarpe** [comment]
Thanks for all the information @saudet. Will have to test that, I suspect it might not play nicely with one underlying dependency, but will have to check. 
Also, will in the next couple of days will start on Android support for core and subsequently I'd like to pair that with Java bindings, so will be able to test more and help on this front.

**Comment by saudet** [comment]
Sounds good! It's easy enough to add Android builds to the presets, as long as the build system of DepthAI plays nice with the NDK. The "hard part" consists in updating the cppbuild.sh file:
https://github.com/bytedeco/javacpp-presets/blob/master/depthai/cppbuild.sh

**Comment by wureka** [comment]
> @wureka here - the following should do:
> 
> ```
> gdb java
> run [name/of/the/java/program] [args if any...]
> ... (after crash)
> bt
> ```
> 
> Change `java` to whatever is the main program that starts the crashing example (in your case that seems to be: `/home/alexji/.sdkman/candidates/java/21.2.0.r11-grl/bin/java`).

@themarpe The problem only happens when running the test code in IntelliJ-IDEA.
If I build the test code to a jar file and run it in terminal console with GraalVM JDK, it works without any problem.

So with the above instruction (`gdb java ....`) , I don't know how to start IntelliJ-IDEA.

**Comment by themarpe** [comment]
@saudet sorry to reuse this thread as a question

I'm wondering if its possible to do "snapshot" builds using `javacpp-present` infrastructure for "development" versions of packages?
I've added (PoC) Android support for DepthAI (`depthai-core` - `xlink_device_search_improvements`), but its not going to be mainlined and released soon yet. Still I'd like to create Android prebuilts /w java bindings to be able to easily distribute and use in an Android Studio project by pointing to your repository for this dependency. 

I saw that I have to fix the build script to target Android as well, but I haven't dug too much into it yet otherwise. 
Any pointers into this direction greatly appreciated:)

**Comment by saudet** [comment]
As long as you're ok to have everything public anyway, sure that's fine.
You'll need to update the workflow file here with "android-arm", etc for
that and that should be pretty much it:
https://github.com/bytedeco/javacpp-presets/tree/master/.github/workflows
Use the one for OpenCV as reference.

**Comment by saudet** [comment]
Ok, I think we can say that all the issues here have been resolved. The signal handler was added in 2.11.0, see https://github.com/luxonis/depthai-core/releases/tag/v2.11.0, so that explains why 2.10.0 works, but not 2.11.0, and I've since disabled that in commit https://github.com/bytedeco/javacpp-presets/commit/5ac61a83a69c05fe7f71e88d6614598101e4b2da. @themarpe Be sure to also disable it in your experimental branch for Android.
